### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SEMANTIC MARKUP]: To-do card list should be written with simple and semantic markup to prevent unintended defects

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#24589]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

- As a screen reader user, I expect cards to be announced as a list so I can keep track on how many exist.
- As a screen reader user, I expect titles within cards to be headings so I can easily identify and navigate to them.
- As a screen reader user, I expect my landmarks list to be uniquely labelled so I can understand the format of the page.
- As a screen reader user, I expect sections to include headings.

## Details

The list of cards on the questionnaire page create several issues that can be remediated by simplifying the markup including:
- The `ol` should not have an `aria-label` (it's already described in the heading) and should have a `role="list"` so it is interpreted as a list on VoiceOver/Safari.
- Using a `header` instead of a `h` will add multiple, unlabelled banner landmarks into a screen reader's landmark list which is undesired.
- Using multiple `section` elements within the card adds false semantics to the markup as it suggests that its contents are a generic section of content within the document. They should always have a `h` as well, which isn't the case for a complete by date (which could be better written as a definition list) and a call to action link (which doesn't need a section). 
- The `aria-label` for the action link is too verbose and does not match the visible label. Because it doesn't match the visible label, if a voice AT user said "Click answer questions", the AT will fail to register it.

https://user-images.githubusercontent.com/14154792/117986395-93c16480-b307-11eb-9195-2535dd78785b.mov

## Acceptance Criteria

- [x] `ol`'s `aria-label` is removed
- [x] `ol` has the attribute `role="list"`
- [x] `header` is replaced with a `h3`
- [x] `section` elements are replaced with `div` elements if they are being used as styling wrappers
- [x] `aria-label` for "Answer questions" is shortened and written as follows: "Answer questions for your visit on January, 18, 2021" (assuming date is the most unique of the options to choose from)

If design allows:

- [x] use a `dl` instead of a `p` and `div`s to organize card content

## Steps to Recreate

Inspect card list on questionnaires page

## Proposed Solution

Simplify the markup as listed in the acceptance criteria.

### Before 
``` html
<ol class="questionnaire-list toDo" aria-label="Your outstanding questionnaires">
  <li class="card">
    <span class="usa-label">not started</span>
    <header>Psychiatry clinic questionnaire</header>
    <p class="appointment-location">for your appointment at TEM MH PSO TRS IND93EH 2, CHEYENNE VA MEDICAL</p>
    <section class="due-details">
      <section class="due-date">
        <p>Complete by</p>
        <p class="vads-u-font-weight--bold">Monday, January 18, 2021</p>
        <p class="vads-u-font-weight--bold">12:00 a.m. PDT</p>
      </section>
    </section>
    <section class="action-link-container">
      <a href="#" class="vads-c-action-link--blue vads-u-padding-left--0" aria-label="Select to start your pre-appointment questionnaire for your primary care visit at CHEYENNE VA MEDICAL on January, 18, 2021">
        <span data-testid="button-text">
          Answer questions</span>
      </a>
    </section>
  </li>
</ol>
```

### Minimum changes in place
``` html
<!-- Remove aria-label from ol and add role="list"-->
<ol class="questionnaire-list toDo" role="list">
  <li class="card">
    <span class="usa-label">not started</span>
    <!-- Change header to heading level 3, there are utility classes if you need to adjust the top margin -->
    <h3 class="vads-u-margin-top--1p5">Psychiatry clinic questionnaire</h3>
    <p class="appointment-location">for your appointment at TEM MH PSO TRS IND93EH 2, CHEYENNE VA MEDICAL</p>
    <!-- Change sections to divs -->
    <div class="due-details">
      <div class="due-date">
        <p>Complete by</p>
        <p class="vads-u-font-weight--bold">Monday, January 18, 2021</p>
        <p class="vads-u-font-weight--bold">12:00 a.m. PDT</p>
      </div>
    </div>
    <div class="action-link-container">
      <!-- Shorten aria label and have it match the visible label -->
      <a href="#" class="vads-c-action-link--blue vads-u-padding-left--0" aria-label="Answer questions for your visit on January, 18, 2021">
        <span data-testid="button-text">
          Answer questions</span>
      </a>
    </div>
  </li>
</ol>
```

### With a definition list (recommended)
``` html
<ol class="questionnaire-list toDo" role="list">
  <li class="card">
    <span class="usa-label">not started</span>
    <h3 class="vads-u-margin-top--1p5">Psychiatry clinic questionnaire</h3>
    <!-- Use a definition list -->
    <dl class="vads-u-margin-bottom--0p5">
      <dt>For your appointment at</dt>
      <dd>Cheyenne VA Medical Center</dd>
      <dt class="vads-u-margin-top--1p5">Complete by</dt>
      <!-- Use the time element -->
      <dd>Saturday <time datetime="2021-07-10T14:00">July 10, 2021 at 2:00 p.m.</time> PDT</dd>
    </dl>
    <div class="action-link-container">
      <a href="#" class="vads-c-action-link--blue vads-u-padding-left--0" aria-label="Answer questions for your visit on January, 18, 2021">
        <span data-testid="button-text">
          Answer questions</span>
      </a>
    </div>
  </li>
</ol>
```

#### Demo of recommended parsing vs. current parsing
Notice that the definition list helps clearly split up sections within the card to screen reader users.

https://user-images.githubusercontent.com/14154792/117993343-a5a60600-b30d-11eb-90e5-df595c31d6c5.mov

#### Demo of recommended card elements list behavior vs. current card elements list behavior

Notice that the heading within the card is now read out within the headings list and navigable to in the recommended version. Note that the current card's markup choice of using a `header` will emerge in the landmarks list which is misleading as this is not a major landmark on the page.
https://user-images.githubusercontent.com/14154792/117993343-a5a60600-b30d-11eb-90e5-df595c31d6c5.mov

## Resources
I recommend the following resources to improve the markup's semantic quality in the future 🙇 

### [MDN: `section`: The Generic Section element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section)
> The HTML `section` element represents a generic standalone section of a document, which doesn't have a more specific semantic element to represent it. 

> Sections should always have a heading, with very few exceptions. If you are only using the element as a styling wrapper, use a `div`. A rule of thumb is that a <section> should logically appear in the outline of a document.

### [Heydon Pickering: Inclusive Cards](https://inclusive-components.design/cards/)
> Let's imagine a basic card. It has an illustration, a 'title', a description, and an attribution. Importantly, it belongs to a list item because — like playing cards — you rarely see one card on its own. 

> Note the heading: like the list markup, headings provide navigation cues to screen reader users. Each card has a heading of the same level — `h2` in this case — because they belong to a flat list hierarchy.

### [MDN: `<dl>`: The Description List element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl)

> The element encloses a list of groups of terms (specified using the `dt` element) and descriptions (provided by `dd` elements). Common uses for this element are to implement a glossary or to display metadata (a list of key-value pairs).

### [MDN: `header`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/header)

> The `header` element is not sectioning content and therefore does not introduce a new section in the outline. That said, a `header` element is intended to usually contain the surrounding section's heading (an h1–h6 element), but this is not required.

### [VA.gov utility classes](https://design.va.gov/utilities/)

> Utilities are helpful, usually single-property classes that can provide more flexibility into a design system at a high specificity. Designers and developers can use them to test new components or build prototypes without committing unnecessary one-time-use selectors into production.

**Comment by mdewey** [comment]
More for awareness, but when implementing the recommended changes, I am getting a linter error when adding `role="list"` to the `<ol>` tag. The error is saying it's redundant. I have to remove it to commit my code. 

<img width="669" alt="Screen Shot 2021-05-14 at 11 37 55 AM" src="https://user-images.githubusercontent.com/1793923/118294547-d40c1880-b4a8-11eb-9878-e1a56a44f208.png">

**Comment by joshkimux** [comment]
@1Copenut , any way we can get past this for the Safari/VO issue that takes away list announcements when the bullet style is removed?

**Comment by 1Copenut** [comment]
@mdewey You can add an ES Lint ignore rule for this line. It'll trigger a bot in the PR, but we can approve it as needed for screen reader usability.

```javascript
// eslint-disable-next-line no-redundant-roles
<ul role="list">
  ...
</ul>
```

**Comment by mdewey** [comment]
Sure thing! @1Copenut .. 

I have the PR here that needs VSP review. https://github.com/department-of-veterans-affairs/vets-website/pull/17221

**Comment by mdewey** [comment]
@joshkimux. This is in staging :-)

**Comment by joshkimux** [comment]
@mdewey great work!! 🥳 
One minor thing to be addressed, the times don't seem to match up here:

<img width="1388" alt="Screen Shot 2021-05-24 at 11 58 14 AM" src="https://user-images.githubusercontent.com/14154792/119374305-6d94b080-bc87-11eb-9a53-0cef40949f86.png">

If getting the `datetime` element to work is difficult, we can pass it through as a normal string to get to launch and save this as a to-do later item.

**Comment by mdewey** [comment]
Great catch @joshkimux! The fix for that is in staging. 

![Screen Shot 2021-05-26 at 10 39 06 AM](https://user-images.githubusercontent.com/1793923/119679772-9f844f00-be0e-11eb-8237-890d20434742.png)

Let me know if you see anything else :-)

**Comment by joshkimux** [comment]
Closing this out 🥳
